pkg/logql/filter.go

```diff
@@ -0,0 +1,287 @@
+package logql
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"regexp/syntax"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// LineFilter is a interface to filter log lines.
+type LineFilter interface {
+	Filter(line []byte) bool
+}
+
+// LineFilterFunc is a syntax sugar for creating line filter from a function
+type LineFilterFunc func(line []byte) bool
+
+func (f LineFilterFunc) Filter(line []byte) bool {
+	return f(line)
+}
+
+type notFilter struct {
+	LineFilter
+}
+
+func (n notFilter) Filter(line []byte) bool {
+	return !n.LineFilter.Filter(line)
+}
+
+// newNotFilter creates a new filter which matches only if the base filter doesn't match.
+// If the base filter is a `or` it will recursively simplify with `and` operations.
+func newNotFilter(base LineFilter) LineFilter {
+	// not(a|b) = not(a) and not(b)
+	if or, ok := base.(orFilter); ok {
+		return newAndFilter(newNotFilter(or.left), newNotFilter(or.right))
+	}
+	return notFilter{LineFilter: base}
+}
+
+type andFilter struct {
+	left  LineFilter
+	right LineFilter
+}
+
+// newAndFilter creates a new filter which matches only if left and right matches.
+func newAndFilter(left LineFilter, right LineFilter) LineFilter {
+	return andFilter{
+		left:  left,
+		right: right,
+	}
+}
+
+func (a andFilter) Filter(line []byte) bool {
+	return a.left.Filter(line) && a.right.Filter(line)
+}
+
+type orFilter struct {
+	left  LineFilter
+	right LineFilter
+}
+
+// newOrFilter creates a new filter which matches only if left or right matches.
+func newOrFilter(left LineFilter, right LineFilter) LineFilter {
+	return orFilter{
+		left:  left,
+		right: right,
+	}
+}
+
+// chainOrFilter is a syntax sugar to chain multiple `or` filters. (1 or many)
+func chainOrFilter(curr, new LineFilter) LineFilter {
+	if curr == nil {
+		return new
+	}
+	return newOrFilter(curr, new)
+}
+
+func (a orFilter) Filter(line []byte) bool {
+	return a.left.Filter(line) || a.right.Filter(line)
+}
+
+type regexpFilter struct {
+	*regexp.Regexp
+}
+
+// newRegexpFilter creates a new line filter for a given regexp.
+// If match is false the filter is the negation of the regexp.
+func newRegexpFilter(re string, match bool) (LineFilter, error) {
+	reg, err := regexp.Compile(re)
+	if err != nil {
+		return nil, err
+	}
+	f := regexpFilter{reg}
+	if match {
+		return f, nil
+	}
+	return newNotFilter(f), nil
+}
+
+func (r regexpFilter) Filter(line []byte) bool {
+	return r.Match(line)
+}
+
+type literalFilter []byte
```

Comment:
Nit: IMO calling this a `containsFilter` would be more descriptive to what it's doing. Or maybe `literalContainsFilter?`